File: internal/router/game.go
package router

import (
	"github.com/elabosak233/cloudsdale/internal/controller"
	"github.com/elabosak233/cloudsdale/internal/model"
	"github.com/gin-gonic/gin"
)

type IGameRouter interface {
	Register()
}

type GameRouter struct {
	router     *gin.RouterGroup
	controller controller.IGameController
}

func NewGameRouter(gameRouter *gin.RouterGroup, gameController controller.IGameController) IGameRouter {
	return &GameRouter{
		router:     gameRouter,
		controller: gameController,
	}
}

func (g *GameRouter) Register() {
	g.router.GET("/", g.PreProcess(), g.controller.Find)
	g.router.POST("/", g.controller.Create)
	g.router.PUT("/:id", g.controller.Update)
	g.router.DELETE("/:id", g.controller.Delete)
	g.router.GET("/:id/challenges", g.controller.FindChallenge)
	g.router.POST("/:id/challenges", g.controller.CreateChallenge)
	g.router.PUT("/:id/challenges/:challenge_id", g.controller.UpdateChallenge)
	g.router.DELETE("/:id/challenges/:challenge_id", g.controller.DeleteChallenge)
	g.router.GET("/:id/teams", g.controller.FindTeam)
	g.router.GET("/:id/teams/:team_id", g.controller.FindTeamByID)
	g.router.POST("/:id/teams", g.controller.CreateTeam)
	g.router.PUT("/:id/teams/:team_id", g.controller.UpdateTeam)
	g.router.DELETE("/:id/teams/:team_id", g.controller.DeleteTeam)
	g.router.GET("/:id/notices", g.controller.FindNotice)
	g.router.POST("/:id/notices", g.controller.CreateNotice)
	g.router.PUT("/:id/notices/:notice_id", g.controller.UpdateNotice)
	g.router.DELETE("/:id/notices/:notice_id", g.controller.DeleteNotice)
	g.router.GET("/:id/broadcast", g.controller.BroadCast)
	g.router.POST("/:id/poster", g.controller.SavePoster)
	g.router.DELETE("/:id/poster", g.controller.DeletePoster)
}

func (g *GameRouter) PreProcess() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		user := ctx.MustGet("user").(*model.User)
		if !(user.Group == "admin") {
			ctx.Set("is_enabled", true)
		}
		ctx.Next()
	}
}

File: src/web/router/game.rs
use crate::util::jwt::Group;
use crate::web::controller;
use crate::web::middleware::auth;
use axum::{
    extract::DefaultBodyLimit,
    middleware::from_fn,
    routing::{delete, get, post, put},
    Router,
};

pub fn router() -> Router {
    return Router::new()
        .route(
            "/",
            get(controller::game::get).layer(from_fn(auth::jwt(Group::User))),
        )
        .route(
            "/",
            post(controller::game::create).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id",
            put(controller::game::update).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id",
            delete(controller::game::delete).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/challenges",
            get(controller::game::get_challenge).layer(from_fn(auth::jwt(Group::User))),
        )
        .route(
            "/:id/challenges",
            post(controller::game::create_challenge).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/challenges/:challenge_id",
            put(controller::game::update_challenge).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/challenges/:challenge_id",
            delete(controller::game::delete_challenge).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/teams",
            get(controller::game::get_team).layer(from_fn(auth::jwt(Group::User))),
        )
        .route(
            "/:id/teams",
            post(controller::game::create_team).layer(from_fn(auth::jwt(Group::User))),
        )
        .route(
            "/:id/teams/:team_id",
            put(controller::game::update_team).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/teams/:team_id",
            delete(controller::game::delete_team).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/notices",
            get(controller::game::get_notice).layer(from_fn(auth::jwt(Group::User))),
        )
        .route(
            "/:id/notices",
            post(controller::game::create_notice).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/notices/:notice_id",
            put(controller::game::update_notice).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/notices/:notice_id",
            delete(controller::game::delete_notice).layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/submissions",
            get(controller::game::get_submission).layer(from_fn(auth::jwt(Group::User))),
        )
        .route("/:id/poster", get(controller::game::get_poster))
        .route(
            "/:id/poster",
            post(controller::game::save_poster)
                .layer(DefaultBodyLimit::max(3 * 1024 * 1024 /* MB */))
                .layer(from_fn(auth::jwt(Group::Admin))),
        )
        .route(
            "/:id/poster/metadata",
            get(controller::game::get_poster_metadata),
        )
        .route(
            "/:id/poster",
            delete(controller::game::delete_poster).layer(from_fn(auth::jwt(Group::Admin))),
        );
}
